### PR TITLE
feat:Add advertisement banners to the forum

### LINKED ISSUE #218: Add advertisement banners to the forum

### PR DESCRIPTION
**Pull request by oranges-eating**
Fix: https://github.com/casbin/casnode/issues/218

An ad slot has been added to the right of the site, and the administrator can modify it in the background

### CODE REVIEW

File: web/src/admin/AdminPoster.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        this.state = {
+            classes: props,
+            message: "",
+            form: {},
+        };
+
+        this.state.url = `/admin/poster`;
+    }
+
+    componentDidMount() {
```
Comment (kocoler) [inline web/src/admin/AdminPoster.js:42]: For the admin page, I think you should read the current advertisement and display it.

File: web/src/admin/AdminPoster.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                        </table>
+                    </div>
+                </div>
+            </div>
+
+        );
+    }
+}
+
+export default withRouter(AdminMember);
```
Comment (kocoler) [inline web/src/admin/AdminPoster.js]: There should be a new line at the end of file.

File: controllers/poster.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2020 The casbin Authors. All Rights Reserved.
```
Comment (kocoler) [inline controllers/poster.go]: Copyright 2021.

File: object/poster.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package object
+
+type Poster struct {
+	Name         string `xorm:"varchar(50) notnull pk" json:"name"`
+	Advertiser   string `xorm:"varchar(40)" json:"advertiser"`
+	Link         string `xorm:"varchar(500)" json:"link"`
+	Picture_link string `xorm:"varchar(500)" json:"picture_link"`
```
Comment (kocoler) [inline object/poster.go]: Use camel case instead of snake case.

File: object/poster.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package object
+
+type Poster struct {
+	Name         string `xorm:"varchar(50) notnull pk" json:"name"`
```
Comment (kocoler) [inline object/poster.go]: There should be an id(autoincr) filed and an deleted(we use soft delete) filed.

File: object/poster.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	if existed {
+		return &poster
+	} else {
+		return nil
+	}
+}
+
+func UpdatePoster(name string, poster Poster) bool {
```
Comment (kocoler) [inline object/poster.go]: If the struct changed, `UpdatePoster` should be changed as well, update `deleted` filed.

File: web/src/admin/AdminPoster.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+import * as MemberBackend from "../backend/MemberBackend.js";
+import * as PosterBackend from "../backend/PosterBackend"
+import * as Setting from "../Setting";
+import PageColumn from "../main/PageColumn";
+import Avatar from "../Avatar";
+import Collapse, { Panel } from "rc-collapse";
+import Select2 from "react-select2-wrapper";
+import i18next from "i18next";
+
+class AdminMember extends React.Component {
```
Comment (kocoler) [inline web/src/admin/AdminPoster.js]: Change to AdminPoster.

**Review by kocoler — CHANGES_REQUESTED** [review]
For each file, there should be a new line at the end of file.

**Comment by hsluoyz** [comment]
@oranges-eating don't close PR frequently. The PR can be updated.

**Comment by hsluoyz** [comment]
@kocoler @Kininaru plz review.

**Comment by hsluoyz** [comment]
@oranges-eating plz fix:

image

File: web/src/App.js [hunk 8]
```diff
@@ -423,6 +424,12 @@ class App extends Component {
             <AdminTab account={this.state.account} />
           </div>
         </Route>
+        <Route exact path="/admin/poster">
+              <div id={pcBrowser ? "Main" : ""}>
+                  {pcBrowser ? <div className="sep20" /> : null}
+                  <AdminPoster account={this.state.account} />
```
Comment (keainye) [inline web/src/App.js]: It seems like that `account` is not used in `AdminPoster`. Can we remove it?

File: object/poster.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
+package object
```
Comment (keainye) [inline object/poster.go:15]: Casbin License

File: web/src/admin/AdminPoster.js [hunk 10]
```diff
@@ -0,0 +1,183 @@
+import React from "react";
```
Comment (keainye) [inline web/src/admin/AdminPoster.js:15]: Casbin license

File: web/src/backend/PosterBackend.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
+// Copyright 2020 The casbin Authors. All Rights Reserved.
```
Comment (keainye) [inline web/src/backend/PosterBackend.js]: 2021

File: web/src/App.js (hunk 8 above)
Comment (oranges-eating) [reply web/src/App.js]: > It seems like that `account` is not used in `AdminPoster`. Can we remove it?

yes,i think

**Review by kocoler — APPROVED** [review]
LGTM

**Review by keainye — APPROVED** [review]
LGTM

**Pull request merged**